curves/bls12_377/src/curves/mod.rs

```diff
@@ -23,15 +21,30 @@ pub use self::{
 #[derive(Clone, Copy)]
 pub struct Config<H: CurveHooks>(PhantomData<fn() -> H>);
 
-pub trait CurveHooks: 'static {
-    fn bls12_377_multi_miller_loop(a: Vec<u8>, b: Vec<u8>) -> Result<Vec<u8>, ()>;
-    fn bls12_377_final_exponentiation(f12: Vec<u8>) -> Result<Vec<u8>, ()>;
-    fn bls12_377_msm_g1(bases: Vec<u8>, scalars: Vec<u8>) -> Result<Vec<u8>, ()>;
+pub trait CurveHooks: 'static + Sized {
+    fn bls12_377_multi_miller_loop(
+        g1: impl Iterator<Item = <Bls12_377<Self> as Pairing>::G1Prepared>,
+        g2: impl Iterator<Item = <Bls12_377<Self> as Pairing>::G2Prepared>,
+    ) -> Result<<Bls12_377<Self> as Pairing>::TargetField, ()>;
+
+    fn bls12_377_final_exponentiation(
+        target: <Bls12_377<Self> as Pairing>::TargetField,
+    ) -> Result<<Bls12_377<Self> as Pairing>::TargetField, ()>;
+
+    fn bls12_377_msm_g1(
+        bases: &[g1::G1SWAffine<Self>],
+        scalars: &[<g1::Config<Self> as CurveConfig>::ScalarField],
+    ) -> Result<g1::G1Projective<Self>, ()>;
+
     fn bls12_377_mul_projective_g1(base: Vec<u8>, scalar: Vec<u8>) -> Result<Vec<u8>, ()>;
+
     fn bls12_377_msm_g2(bases: Vec<u8>, scalars: Vec<u8>) -> Result<Vec<u8>, ()>;
+
     fn bls12_377_mul_projective_g2(base: Vec<u8>, scalar: Vec<u8>) -> Result<Vec<u8>, ()>;
 }
```

Comment:
Is this the dependency injection trait?  I'm surprised it lives in the curve, not in the model.

Yes I'd think the dependency injection trait should mirror the outer call if possible.

Reply:
There is one *Hook* trait for each curve to implement the selected ops.

These traits live here because are used as generic params to the `Config` structs, which are defined in the same path in the upstream arkworks as well.